### PR TITLE
Improve object selection

### LINKED ISSUE #56: No difference between `<a-p>` and `<i-p>`
Selecting a paragraph with `<a-p>` and `<i-p>` doesn't make a difference. `<a-p>` should select empty lines after the paragraph.

### LINKED ISSUE #57: Inconsistent behaviour of `<a-p>` and `<i-p>` on last paragraph of file
Using `<a-p>` and `<i-p>` on the last paragraph of a file doesn't select the whole paragraph, it only selected upwards, till the last newline. Only happens when the last line of the file is not an empty line!

### PR DESCRIPTION
**Pull request by pomdtr**
Close #56 
Close #57

This pull request fix the handling of paragraphs, by only using empty lines as delimiters.

### CODE REVIEW

**Comment by 71** [comment]
This doesn't seem to work. For instance, `a-a p` extends the current selection, and doesn't correctly the next paragraph on first invocation.

**Comment by pomdtr** [comment]
Sorry, I don't understand your point. Kakoune also extend selection on first invocation ?

**Comment by 71** [comment]
Sorry, my comment was pretty confusing. I noticed two different issues:
1. It extends the selection, even on multiple invocations.
2. On first invocation, it does not select a paragraph but only the first few characters of the line.

I tried it on a README file, so not on code.

**Comment by pomdtr** [comment]
1. Ok, I see the problem. It is hard to fix while using the original signature of the function since only the position of the start/end is accessible. This is also the current behavior in dance 0.2.1.
2. Can't reproduce. Can you give me a text snippet in order to fix the issue?

**Comment by pomdtr** [comment]
79e0736 fix the first point. It also corrects some dance incoherent behavior.
See #70 description.

**Comment by pomdtr** [comment]
So i order to fix the problem, i did simplify the code of the performObjectSelect function. I tested all impacted command and they seems to work. Also, invert selection (ex: `[p`) now correctly set the cursor in front of the selection.

File: src/commands/selectObject.ts [hunk 1]
```diff
@@ -73,79 +73,79 @@ function findObjectWithChars(text: TextBuffer, backwards: boolean, inner: boolea
...
-    if (c === '(') {
-      balance++
-    } else if (c === ')') {
-      balance--
-    } else if (c === undefined) {
-      return text.position(i + 1)!
-    } else if (balance !== 0) {
-      // Nop.
+  // Go to the end of the last sentence
+  while(text.char(i) == ' '){
```
Comment (71) [inline src/commands/selectObject.ts]:
```suggestion
  while (text.char(i) == ' '){
```

File: src/commands/selectObject.ts [hunk 2]
```diff
@@ -73,79 +73,79 @@ function findObjectWithChars(text: TextBuffer, backwards: boolean, inner: boolea
...
-    } else if (c === ')') {
-      balance--
-    } else if (c === undefined) {
-      return text.position(i + 1)!
-    } else if (balance !== 0) {
-      // Nop.
+  let i = 0
+  for (let c = text.char(i);; c = text.char(i++)) {
+    if (c === undefined) {
+      return text.position(i-1)!
```
Comment (71) [inline src/commands/selectObject.ts]:
```suggestion
      return text.position(i - 1)!
```

File: src/commands/selectObject.ts [hunk 3]
```diff
@@ -73,79 +73,79 @@ function findObjectWithChars(text: TextBuffer, backwards: boolean, inner: boolea
...
     } else if (c === '\n' && text.char(i + 1) === '\n') {
       return text.position(i)!
     } else if (c === '.' || c === ';' || c === '!' || c === '?') {
-      return text.position(i + (inner ? 1 : 0))!
+      if (inner)
+        return text.position(i)!
+      for (i++;; c = text.char(i++)) {
+        if (c=== undefined || isAlphaWord(c)) {
```
Comment (71) [inline src/commands/selectObject.ts]:
```suggestion
        if (c === undefined || isAlphaWord(c)) {
```

File: src/commands/selectObject.ts [hunk 4]
```diff
@@ -73,79 +73,79 @@ function findObjectWithChars(text: TextBuffer, backwards: boolean, inner: boolea
...
+          return text.position(i-1)!
```
Comment (71) [inline src/commands/selectObject.ts]:
```suggestion
          return text.position(i - 1)!
```

File: src/commands/selectObject.ts [hunk 5]
```diff
@@ -373,92 +373,69 @@ function findObjectEnd(text: TextBuffer, type: number, inner: boolean): vscode.P
...
   lastObjectSelectOperation = [inner, type, extend, toStart, toEnd]
 
   editor.selections = editor.selections.map(selection => {
-    let start = selection.start
-    let end = selection.end
-    let inInfiniteLooop = false
+    let start = selection.active
+    if (shift) {
+      if (selection.active.character !== 0)
+        start = start.translate({characterDelta: -1})
```
Comment (71) [inline src/commands/selectObject.ts]:
```suggestion
        start = start.translate({ characterDelta: -1 })
```

File: src/commands/selectObject.ts [hunk 6]
```diff
@@ -73,79 +73,79 @@ function findObjectWithChars(text: TextBuffer, backwards: boolean, inner: boolea
...
-      balance--
-    } else if (c === undefined) {
-      return text.position(i + 1)!
-    } else if (balance !== 0) {
-      // Nop.
-    } else if (c === '\n' && text.char(i - 1) === '\n') {
-      return text.position(i + 1)!
+function findParagraphStart(text: TextBuffer, inner: boolean): vscode.Position {
+  let lineNumber = text.line.lineNumber
+  if (!inner) {
```
Comment (71) [inline src/commands/selectObject.ts]:
```suggestion
  let lineNumber = text.line.lineNumber

  if (!inner) {
```

File: src/commands/selectObject.ts [hunk 7]
```diff
@@ -73,79 +73,79 @@ function findObjectWithChars(text: TextBuffer, backwards: boolean, inner: boolea
...
-      return text.position(i + 1)!
-    } else if (balance !== 0) {
-      // Nop.
-    } else if (c === '\n' && text.char(i + 1) === '\n') {
-      return text.position(i)!
-    }
+  if (!inner) {
+    // Select all empty lines below the paragraph
+    while (lineNumber <= lastLine && text.doc.lineAt(lineNumber).isEmptyOrWhitespace)
+    lineNumber++
```
Comment (71) [inline src/commands/selectObject.ts]:
```suggestion
      lineNumber++
```

Comment (71) [reply src/commands/selectObject.ts]: For consistency, you should either add braces here or remove the braces above.

File: src/commands/selectObject.ts [hunk 8]
```diff
@@ -73,79 +73,79 @@ function findObjectWithChars(text: TextBuffer, backwards: boolean, inner: boolea
...
+  for (let c = text.char(i);; c = text.char(i++)) {
+    if (c === undefined) {
+      return text.position(i-1)!
     } else if (c === '\n' && text.char(i + 1) === '\n') {
       return text.position(i)!
     } else if (c === '.' || c === ';' || c === '!' || c === '?') {
-      return text.position(i + (inner ? 1 : 0))!
+      if (inner)
+        return text.position(i)!
+      for (i++;; c = text.char(i++)) {
```
Comment (71) [inline src/commands/selectObject.ts]:
```suggestion
      for (i += 1;; c = text.char(i++)) {
```

**Comment by 71** [comment]
Please don't update this PR. Due to the upcoming refactor, a large part of `selectObject.ts` must be changed, so I'll try to directly implement your changes with the new API (so your work won't be lost). Thank you again for your work!

**Pull request closed without merging**